web/azureschandlers.go

```
@@ -52,22 +61,81 @@ func (ac *AppContext) loginOAuthAzureSC(w http.ResponseWriter, r *http.Request)
 
 	http.Redirect(w, r, fmt.Sprintf("/azuresc-details?t=%s&u=%s", tenant, id), http.StatusFound)
 }
+func getTenant(subid string) string {
+	SubUrl := fmt.Sprintf("https://management.azure.com/subscriptions/%s?api-version=2015-01-01", subid)
+	resp, _ := http.Get(SubUrl)
+
+	defer resp.Body.Close()
+	www := resp.Header.Get("WWW-Authenticate")
+	logrus.Warnf(www)
+	if www != "" {
+		re, _ := regexp.Compile("/[a-zA-Z0-9\\-]*\"")
+		match := re.FindString(www)
+		match = strings.TrimLeft(match, "/")
+		match = strings.TrimRight(match, "\"")
+		return match
+	} else {
+		logrus.Warnf("could not find subscription id with [%s]", subid)
+		return ""
+	}
+}
+
+func getTokens(tenant string, code string) (string, string, string) {
+	BaseUrl := fmt.Sprintf("https://login.microsoftonline.com/%s/OAuth2/token", tenant)
+	body := url.Values{
+		"client_id":     {conf.Options.AzureSC.ClientID},
+		"client_secret": {conf.Options.AzureSC.ClientSecret},
+		"grant_type":    {"authorization_code"},
+		"redirect_uri":  {"https://ec2-63-35-157-116.eu-west-1.compute.amazonaws.com/azuresc"},
+		"code":          {code},
+	}
+
+	r, err := http.Post(BaseUrl, "application/x-www-form-urlencoded", bytes.NewBufferString(body.Encode()))
+	if err != nil {
+		return "", "", ""
+	}
+	defer r.Body.Close()
+
+	var result map[string]interface{}
+
+	json.NewDecoder(r.Body).Decode(&result)
+
+	refreshToken := result["refresh_token"].(string)
+	accessToken := result["access_token"].(string)
+	expiresOn := result["expires_on"].(string)
+
+	if len(refreshToken) == 0 || len(accessToken) == 0 || len(expiresOn) == 0 {
+		return "", "", ""
```

We don't want to return an error on this case?

---

its get checked in the function who's calling it